modules/ui/tag_reference.js

```diff
@@ -10,53 +7,61 @@ import { t } from '../util/locale';
 import { utilDetect } from '../util/detect';
 import { services } from '../services';
 import { svgIcon } from '../svg';
+import { utilQsString } from '../util';
 
 
 export function uiTagReference(tag) {
-    var taginfo = services.taginfo;
+    var wikibase = services.osmWikibase;
     var tagReference = {};
 
     var _button = d3_select(null);
     var _body = d3_select(null);
     var _loaded;
     var _showing;
 
-
+    /**
+     * @returns {{itemTitle: String, description: String, image: String|null}|null}
+     **/
     function findLocal(data) {
-        var locale = utilDetect().locale.toLowerCase();
-        var localized;
-
-        if (locale !== 'pt-br') {  // see #3776, prefer 'pt' over 'pt-br'
-            localized = _find(data, function(d) {
-                return d.lang.toLowerCase() === locale;
-            });
-            if (localized) return localized;
-        }
-
-        // try the non-regional version of a language, like
-        // 'en' if the language is 'en-US'
-        if (locale.indexOf('-') !== -1) {
-            var first = locale.split('-')[0];
-            localized = _find(data, function(d) {
-                return d.lang.toLowerCase() === first;
-            });
-            if (localized) return localized;
+        var entity = data.tag || data.key;
+        if (!entity) return null;
+
+        var result = {
+            title: entity.title,
+            description: wikibase.getDescription(entity),
+        };
+
+        if (entity.claims) {
+            var langCode = utilDetect().locale.toLowerCase();
+            var url;
+            var image = wikibase.claimToValue(entity, 'P4', langCode);
+            if (image) {
+                url = 'https://commons.wikimedia.org/w/index.php';
+            } else {
+                image = wikibase.claimToValue(entity, 'P28', langCode);
+                if (image) {
+                    url = 'https://wiki.openstreetmap.org/w/index.php';
+                }
+            }
+            if (image) {
+                result.image = {
+                    url: url,
+                    title: 'Special:Redirect/file/' + image
```

Comment:
Special:Redirect downloads the full image, which can be [huge](https://wiki.openstreetmap.org/wiki/Item:Q145#P4) for such a small piece of screen real estate.

[This MediaWiki API query](https://wiki.openstreetmap.org/wiki/Special:ApiSandbox#action=query&format=json&prop=imageinfo&titles=File%3ADollar%20General%2C%20Glen%20St.%20Mary.JPG&iiprop=url|size&iiurlwidth=100) gets the 100-pixel-wide thumbnail of an image given the image name. A `maxage` parameter would also make sense, since we don’t expect the images to change very frequently.

Reply:
@1ec5 this is not correct, run these commands:
```
url -L 'https://en.wikipedia.org/w/index.php?title=Special:Redirect/file/File:Dollar_General,_Glen_St._Mary.JPG&width=100&height=100' > small.jpg
curl -L 'https://en.wikipedia.org/w/index.php?title=Special:Redirect/file/File:Dollar_General,_Glen_St._Mary.JPG' > large.jpg
ls -al small.jpg large.jpg
```
small.jpg is 2749 bytes

Reply:
Ah, I missed that the `width` and `height` parameters were being tacked onto the end. As long as using Special:Redirect this way doesn’t raise server-side performance issues like [thumb.php](https://meta.wikimedia.org/wiki/Talk:Www.wikipedia.org_template/2006#thumb.php), then this is much simpler than what I suggested. 🙂

Reply:
Heh, that was a while ago :))   Redirect in theory should just [generate a URL](https://github.com/wikimedia/mediawiki/blob/0e56ff447bbb74df007f2573725baaa61ed2745c/includes/specials/SpecialRedirect.php#L111) according to MediaWiki rules, and redirect to the actual image store - which in turn will handle the missing size to dynamically scale it.